lib/obj.js

```diff
@@ -1,365 +1,479 @@
+var normalPattern = /vn( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                     // vn float float float
+var uvPattern = /vt( +[\d|\.|\+|\-|e|E]+)( +[\d|\.|\+|\-|e|E]+)/;                                                                               // vt float float
+var facePattern1 = /f( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)\/?( +-?\d+)?\/?/;                                                                        // f vertex vertex vertex ...
+var facePattern2 = /f( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)( +(-?\d+)\/(-?\d+)\/?)?/;                            // f vertex/uv vertex/uv vertex/uv ...
+var facePattern3 = /f( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))( +(-?\d+)\/(-?\d+)\/(-?\d+))?/;    // f vertex/uv/normal vertex/uv/normal vertex/uv/normal ...
+var facePattern4 = /f( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))( +(-?\d+)\/\/(-?\d+))?/;                                // f vertex//normal vertex//normal vertex//normal ...
+
+/**
+ * Parse an obj file.
+ *
+ * @param {String} objPath Path to the obj file.
+ * @param {Object} [options] An object with the following properties:
+ * @param {Boolean} [options.hasTransparency=false] Do a more exhaustive check for texture transparency by looking at the alpha channel of each pixel.
+ * @param {Boolean} [options.secure=false] Prevent the converter from reading image or mtl files outside of the input obj directory.
+ * @param {Boolean} [options.logger] A callback function for handling logged messages. Defaults to console.log.
+ * @returns {Promise} A promise resolving to the obj data.
+ * @exception {RuntimeError} The file does not have any geometry information in it.
+ *
+ * @private
+ */
+function loadObj(objPath, options) {
+    // The defaults are set in convert as well, this just helps with testing loadObj individually
+    options = combine(options, {
+        hasTransparency : false,
+        secure : false,
+        logger : function(message) {
+            console.log(message);
         }
+    });
+
+    // Global store of vertex attributes listed in the obj file
+    var positions = new ArrayStorage(ComponentDatatype.FLOAT);
+    var normals = new ArrayStorage(ComponentDatatype.FLOAT);
+    var uvs = new ArrayStorage(ComponentDatatype.FLOAT);
+
+    // The current node, mesh, and primitive
+    var node;
+    var mesh;
+    var primitive;
+
+    // All nodes seen in the obj
+    var nodes = [];
+
+    // Used to build the indices. The vertex cache is unique to each mesh.
+    var vertexCache = {};
+    var vertexCacheLimit = 1000000;
+    var vertexCacheCount = 0;
+    var vertexCount = 0;
+
+    // All mtl paths seen in the obj
+    var mtlPaths = [];
+
+    function getName(name) {
```

Comment:
It is organized this way so the top variables are accessible to the local functions, which I felt was cleaner than passing around a state object to each function. I'm open to changes though, but probably in another PR.